src/v/features/migrators/cloud_storage_config.cc

```diff
@@ -0,0 +1,111 @@
+// Copyright 2022 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#include "features/migrators/cloud_storage_config.h"
+
+#include "cluster/controller.h"
+#include "cluster/feature_manager.h"
+#include "cluster/topics_frontend.h"
+#include "features/feature_migrator.h"
+#include "features/feature_table.h"
+#include "features/logger.h"
+#include "ssx/future-util.h"
+#include "vlog.h"
+
+#include <seastar/core/sleep.hh>
+
+#include <type_traits>
+
+using namespace std::chrono_literals;
+
+namespace features::migrators {
+
+template<typename T>
+cluster::property_update<T> make_property_set(const T& v) {
+    return cluster::property_update<T>(
+      v, cluster::incremental_update_operation::set);
+}
+
+ss::future<> cloud_storage_config::do_mutate() {
+    cluster::topic_table& topic_table = _controller.get_topics_state().local();
+
+    vlog(featureslog.info, "Checking for topic configs to update...");
+    size_t update_count = 0;
+    for (const auto& i : topic_table.topics_map()) {
+        auto& topic_conf = i.second.metadata.get_configuration();
+
+        auto& props = topic_conf.properties;
+
+        bool si_enabled
+          = (props.shadow_indexing
+             && props.shadow_indexing
+                  != model::shadow_indexing_mode::disabled)
+            || ((!props.shadow_indexing || props.shadow_indexing == model::shadow_indexing_mode::disabled) && config::shard_local_cfg().cloud_storage_enable_remote_write());
```

Comment:
I am wondering if we should make migration dependent on current settings, what will happen if remote write is disabled during upgrade and then enabled back ?

Reply:
We're probably going to change the behavior of cluster configs for cloud storage like this, before releasing, so that the issue of subsequent changes doesn't exist for those: https://github.com/redpanda-data/redpanda/issues/6917

If someone specifically switched tiered storage off on one topic before they upgrade, then there isn't much we can do: even if we remembered "this is a legacy topic, switch the retention settings when they enable tiered storage", it wouldn't really be correct, because we don't know if when they set those retention settings their intention was to control total or local retention: the best guess we can make is to respect the current enablement of tiered storage.

There will be some communication to known tiered storage users ahead of the upgrade to explain the change.